### PR TITLE
feat: allow spectral rules to be defined in the validaterc

### PR DESCRIPTION
**Pull request by rmkeezer**
ref: https://github.ibm.com/arf/planning-sdk-squad/issues/2170
followup of: https://github.com/IBM/openapi-validator/pull/195

These changes allow spectral rules to have their severity level changed in the `.validatrc` and allow custom rules defined in a `.spectral.yaml` file to not be overridden by the `.defaultsForSpectral.yaml`. Although the `.defaultsForSpectral.yaml` will still override the `.spectral.yaml` severity level of any existing rule `.defaultsForSpectral.yaml` defines a severity for, including the extended rules.

### CODE REVIEW

File: src/cli-validator/utils/processConfiguration.js
Comment (rmkeezer) [inline src/cli-validator/utils/processConfiguration.js:46]: This entire file is just indention changes with a single if statement, not sure why the github diff shows it so poorly

File: src/spectral/rulesets/.defaultsForSpectral.yaml
Comment (jorge-ibm) [inline src/spectral/rulesets/.defaultsForSpectral.yaml:3]: Surprised spectral didn't complain about that

File: src/spectral/utils/spectral-validator.js
Comment (jorge-ibm) [inline src/spectral/utils/spectral-validator.js:73]: Might be worth trying to see if you can achieve the same merging by simply calling ` await spectral.loadRuleset([spectralRuleset, defaultSpectralRulesetURI]);`

Comment (rmkeezer) [reply src/spectral/utils/spectral-validator.js]: Yup this does achieve the same, I wasn't confident it did at first but it seems to be working as expect! I updated the PR with this

**Review by jorge-ibm — CHANGES_REQUESTED** [review]
It's looking better, just a couple of more things that need to be changed. Also, if possible, some unit tests to test the validaterc changes would be helpful.

File: src/.defaultsForValidator.js
Comment (rmkeezer) [reply src/.defaultsForValidator.js:110]: Since i'm calling `getConfigObject` if there is no spectral section in the validaterc, it will take the rules from the spectral rules from `.defaultsForValidator.js` and these changes also allow the `init` command to generate the new spectral section

File: src/spectral/utils/spectral-validator.js
Comment (rmkeezer) [reply src/spectral/utils/spectral-validator.js:67]: I dont think there is a problem with loading a ruleset twice since it won't create any duplicates. Currently it will load the user config and the `.defaultsForSpectral.yaml` config or the `.defaultsForSpectral.yaml` config twice, how should this be changed? I can change it to load the `.defaultsForSpectral.yaml` first, should I make any other changes?

Comment (jorge-ibm) [reply src/spectral/utils/spectral-validator.js:67]: Okay. Yes, I think we should load the ``.defaultsForSpectral.yaml `` first

Comment (rmkeezer) [reply src/spectral/utils/spectral-validator.js]: Ok i updated it to load the defaults first

**Comment by jorge-ibm** [comment]
@rmkeezer I'm not sure if my suggestion to use ``loadRuleset`` vs ``mergeRules`` broke something, but I pulled your fork locally and wasn't able to overwrite the severity level of our rules using the validaterc file.

**Comment by rmkeezer** [comment]
@jorge-ibm Ah, apparently `spectral.mergeRules` does not work but using the `mergeRules` function from the spectral ruleset does. I've updated the PR with that change, it should now correctly override the severity again.

**Comment by jorge-ibm** [comment]
Looks like everything is working as expected now, though I still think some tests would be helpful. The only thing I think should be fixed before merging is this duplicate message displaying whenever the validaterc file isn't present:
```
JorgeWoComputer:openapi-validator-rkeezer jrangel$ lint-openapi -p test/spectral/mockFiles/oas3/enabled-rules.yml 
[Warning] No .validaterc file found. The validator will run in default mode.
To configure the validator, create a .validaterc file.
[Warning] No .validaterc file found. The validator will run in default mode.
To configure the validator, create a .validaterc file.
```

I'm assuming we want to add documentation on how to enable/edit our spectral rules from the validaterc file. @dpopp07 what are your thoughts?

**Review by jorge-ibm — CHANGES_REQUESTED** [review]
Sorry for requesting another set of changes, but I think this PR is really close.

**Review by jorge-ibm — APPROVED** [review]
LGTM. I'll approve but we should hold off on merging until @dpopp07 or @mkistler review it as well

File: src/spectral/utils/spectral-validator.js
Comment (dpopp07) [inline src/spectral/utils/spectral-validator.js:51]: Is there anywhere `setup` would be called without specifying this value? It looks like the one place we call it, we provide the value. So this default seems redundant. If it can be called in other places, that's a different story!

File: src/spectral/utils/spectral-validator.js
Comment (dpopp07) [inline src/spectral/utils/spectral-validator.js:68]: So this handles the case in which users define their own configuration with the Spectral format? But we don't let them override our default Spectral rules with that format?

I think this is what we want but as someone not very familiar with Spectral, I find it a little confusing. I appreciate the comments here, they are helpful, but will you add a couple more to provide context for why this is there?

**Review by dpopp07 — CHANGES_REQUESTED** [review]
This looks good! I left a few minor comments but the primary reason I am "requesting changes" is because we need to have some tests for this behavior to ensure it does what we expect it to and to prevent regressions as we keep working on the Spectral stuff

File: src/spectral/utils/spectral-validator.js
Comment (dpopp07) [reply src/spectral/utils/spectral-validator.js:51]: Okay, I found it 🙂 This happens [on this line](https://github.com/IBM/openapi-validator/blob/master/src/lib/index.js#L18) but we have the default value there as a variable - so I think it would be better to pass it in that let it be potentially redefined. This should be addressed before approval

File: src/spectral/utils/spectral-validator.js
Comment (jorge-ibm) [reply src/spectral/utils/spectral-validator.js:68]: Matthew correct me if I'm wrong but with this PR this is what the user can do now:
1. They can override the severity level and disable/enable our spectral rules from the validaterc file.
2. They can _add_ any new custom rules to the validator using a custom rule set. Even if they don't include our spectral rules in their ruleset file, our spectral rules will still be loaded.
3. If a user wants to override our spectral rules using their ruleset file, they would have to:
   - remove that spectral rule from the validaterc file
   - add a rule with the same name to their ruleset file

Comment (rmkeezer) [reply src/spectral/utils/spectral-validator.js:69]: 1. Correct
2. Correct
3. Correct, but I realized that the omitting the spectral rule from the validaterc file did _not_ remove it and still overrode the severity the user would specify.

This was due to the config.get() function getting the defaultObject when there was no validaterc found, this default object specified the same rules as the default validaterc and would take priority. So I added a new flag to prevent this defaultObject from overriding it so it acts as you should expect it to like 3.

File: package.json
Comment (rmkeezer) [inline package.json:12]: I added tests to check reading the default config .validaterc, and the user config .spectral.yml, but to test them I had the test create a temporary config file.

However, this interferes with other tests when run in parrallel, so I have this test file running sequentially after the regular tests.

Is there a better solution for tests like these or is it fine to run this file sequentially?

Comment (jorge-ibm) [reply package.json:12]: Maybe try to spy on the fs package? https://stackoverflow.com/questions/50066138/mock-fs-function-with-jest . Not sure, @dpopp07 might have more experience doing something like that.

File: test/spectral/tests/spectral-validator.test.js
Comment (dpopp07) [inline test/spectral/tests/spectral-validator.test.js:116]: My only question on the tests is - is there a way to just look at what rules spectral is ready to use after it's defined? Then we could just call the setup function and make sure it does what we want it to. I think that would be much simpler.

That said, I do like the idea of having an end-to-end test to verify the behavior we're looking for in a real life scenario. We can discuss later

**Review by dpopp07 — CHANGES_REQUESTED** [review]
Getting better but I think there are some simplifications to be made that I did not notice before, specifically around reading the config object. Let me know what you think

**Comment by rmkeezer** [comment]
I updated the PR with @dpopp07 's suggestions for using the defaultObject instead of calling config.get() inside the spectral run. This worked well! I also updated the unit test to using jest.spyOn to replace the file URL and the returned config file with their respective spectral changes. So the test does not need to run sequentially anymore!

File: test/spectral/tests/spectral-validator.test.js
Comment (rmkeezer) [reply test/spectral/tests/spectral-validator.test.js]: I was able to spyOn the config files so the end-to-end test isn't as troublesome anymore. For non-end-to-end tests, it's a bit difficult with spectral since creating a ruleset needs a file first and add/modifying rules is a bit messy using spectral's functions.

File: src/spectral/utils/spectral-validator.js
Comment (dpopp07) [inline src/spectral/utils/spectral-validator.js:71]:
```suggestion
  // The validaterc rules will take precendence in the case of duplicate rules
```

File: src/cli-validator/utils/processConfiguration.js
Comment (dpopp07) [inline src/cli-validator/utils/processConfiguration.js:13]: I think this makes sense but I was confused at first - maybe just add a bit more info in the comment like 
```suggestion
// Clear spectral rules from defaultObject, only use the explicit values in the validaterc file
// the default values for spectral come from another file
```

**Review by jorge-ibm — COMMENTED** [review]
Looks really good! My last recommendation would be to add at least 1 test case that uses the ``commandLineValidator`` vs the ``inCodeValidator`` to invoke the validator.

**Review by dpopp07 — APPROVED** [review]
Looks good! I left some extremely minor comments to address before merging but I think it's good otherwise

**Review by jorge-ibm — APPROVED** [review]

File: test/spectral/tests/spectral-validator.test.js
Comment (rmkeezer) [reply test/spectral/tests/spectral-validator.test.js:268]: Yup, there is a warning for `Definition was declared but never used in document`, i'm clarifying the comment for this

**Review by mkistler — APPROVED** [review]
Looks good! 👍

**Pull request merged**
